src/sign.rs

```diff
@@ -16,6 +16,80 @@ pub struct SpsEqSignature<E: PairingEngine> {
     pub Yp: E::G2Projective,
 }
 
+impl<E: PairingEngine> SpsEqSignature<E> {
+    /// Mutably changes the representation of the signature and a message. The
+    /// function does not make assumptions with regards to the relation between
+    /// the message and the signature (ie. signature may not correspond to the
+    /// message)
+    pub fn change_repr<R>(
+        &mut self,
+        message: &Vec<E::G1Projective>,
+        rng: &mut R,
+    ) -> Vec<E::G1Projective>
+    where
+        R: Rng + CryptoRng,
+    {
+        let rnd_f = E::Fr::rand(rng);
+        let rnd_u = E::Fr::rand(rng);
+
+        let rnd_signature = SpsEqSignature::<E>::rnd_signature(&self, rnd_u);
+        self.Z = rnd_signature.Z;
+        self.Y = rnd_signature.Y;
+        self.Yp = rnd_signature.Yp;
+
+        SpsEqSignature::<E>::rnd_message(message, rnd_f)
+    }
+
+    /// Generates a new representation of the signature and message, and returns
+    /// the new representation of the signature and the message; The function
+    /// does not make assumptions with regards to relation between the message
+    /// and the signature (ie. signature may not correspond to the message)
+    pub fn generate_new_repr<R>(
+        self,
+        message: &Vec<E::G1Projective>,
+        rng: &mut R,
+    ) -> (SpsEqSignature<E>, Vec<E::G1Projective>)
+    where
+        R: Rng + CryptoRng,
+    {
+        let rnd_f = E::Fr::rand(rng);
+        let rnd_u = E::Fr::rand(rng);
+
+        let rnd_signature = SpsEqSignature::<E>::rnd_signature(&self, rnd_u);
+        let rnd_message = SpsEqSignature::<E>::rnd_message(message, rnd_f);
+
+        (rnd_signature, rnd_message)
+    }
+
+    fn rnd_message(message: &Vec<E::G1Projective>, rnd_f: E::Fr) -> Vec<E::G1Projective> {
+        let rnd_msg = message
+            .clone()
+            .into_iter()
+            .map(|mut g| {
+                g *= rnd_f;
+                g
+            })
+            .collect();
+
+        rnd_msg
+    }
+
+    fn rnd_signature(signature: &SpsEqSignature<E>, rnd_u: E::Fr) -> SpsEqSignature<E> {
```

Comment:
There are a couple reasons to define `rnd_signature()` as function and not as a type method. The main reason is because we want to use it in `generate_new_repr() too`, which returns a new representation of the signature (and does not mutate the object's state upon which the method `generate_new_repr()` was called). Also, we don't want to expose that to the public, so we can see `rnd_signature()` as a [pure-ish](https://en.wikipedia.org/wiki/Pure_function) helper function that does not mutate any state.

Reply:
SGTM